### ISSUE TITLE
Cannot find attribute `serde` in this scope.

### ISSUE THREAD

**Issue by Yuhanun** [stepancheg/rust-protobuf#526]
Hello,

I'm working on my build script, however, it doesn't seem to be able to compile the generated .rs files with the serde json features.

Cargo.toml: 

```toml
[package]
name = "concrete_proto"
version = "0.1.0"
authors = ["..."]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
protobuf = { git = "https://github.com/stepancheg/rust-protobuf.git", features = ["with-serde"] }
serde = "1.0.57"

[build-dependencies]
protoc-rust = { git = "https://github.com/stepancheg/rust-protobuf.git" }
```

build.rs:
```rs
fn generate_code<'a>(name: &'a str, include: &'a str) {
    protoc_rust::Codegen::new()
        .out_dir("src")
        .include(include)
        .input(name)
        .customize(protoc_rust::Customize {
            serde_derive: Some(true),
            ..Default::default()
        })
        .run()
        .unwrap();
}

fn generate_all(dir: &str, out: &mut Vec<String>) -> std::io::Result<()> {
    for entry in std::fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_dir() {
            generate_all(path.to_str().unwrap(), out).unwrap();
            continue;
        }
        match path.extension() {
            Some(ext) => {
                if ext == "proto" {
                    out.push(format!("pub mod {};", path.file_stem().unwrap().to_str().unwrap()));
                    generate_code(path.to_str().unwrap(), dir);
                }
            }
            None => {}
        }
    }

    Ok(())
}

fn main() {
    let mut out = vec![
        String::from("pub use serde::*;")
    ];
    generate_all("proto", &mut out).unwrap();
    std::fs::remove_file("src/lib.rs").unwrap();
    std::fs::write(
        "src/lib.rs",
        out.into_iter()
            .map(|e| format!("{}", e))
            .collect::<Vec<String>>()
            .join("\n"),
    )
    .unwrap();
}
```

Proto file: 

```proto
syntax = "proto3";

package concrete.protobuf;

message Channel {
    uint64 id = 1;
    string name = 2;
    string description = 3;
}
```

How come this errors with the following:

```rs
 ~/c/backend   master ±  cargo run
   Compiling concrete_proto v0.1.0 (/home/john/concrete/backend/concrete_proto)
error: cannot find derive macro `Serialize` in this scope
  --> concrete_proto/src/user.rs:29:10
   |
29 | #[derive(Serialize, Deserialize)]
   |          ^^^^^^^^^

error: cannot find derive macro `Deserialize` in this scope
  --> concrete_proto/src/user.rs:29:21
   |
29 | #[derive(Serialize, Deserialize)]
   |                     ^^^^^^^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/user.rs:37:7
   |
37 |     #[serde(skip)]
   |       ^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/user.rs:39:7
   |
39 |     #[serde(skip)]
   |       ^^^^^

error: cannot find derive macro `Serialize` in this scope
  --> concrete_proto/src/message.rs:29:10
   |
29 | #[derive(Serialize, Deserialize)]
   |          ^^^^^^^^^

error: cannot find derive macro `Deserialize` in this scope
  --> concrete_proto/src/message.rs:29:21
   |
29 | #[derive(Serialize, Deserialize)]
   |                     ^^^^^^^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/message.rs:38:7
   |
38 |     #[serde(skip)]
   |       ^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/message.rs:40:7
   |
40 |     #[serde(skip)]
   |       ^^^^^

error: cannot find derive macro `Serialize` in this scope
  --> concrete_proto/src/channel.rs:29:10
   |
29 | #[derive(Serialize, Deserialize)]
   |          ^^^^^^^^^

error: cannot find derive macro `Deserialize` in this scope
  --> concrete_proto/src/channel.rs:29:21
   |
29 | #[derive(Serialize, Deserialize)]
   |                     ^^^^^^^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/channel.rs:36:7
   |
36 |     #[serde(skip)]
   |       ^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/channel.rs:38:7
   |
38 |     #[serde(skip)]
   |       ^^^^^

error: cannot find derive macro `Serialize` in this scope
  --> concrete_proto/src/server.rs:29:10
   |
29 | #[derive(Serialize, Deserialize)]
   |          ^^^^^^^^^

error: cannot find derive macro `Deserialize` in this scope
  --> concrete_proto/src/server.rs:29:21
   |
29 | #[derive(Serialize, Deserialize)]
   |                     ^^^^^^^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/server.rs:37:7
   |
37 |     #[serde(skip)]
   |       ^^^^^

error: cannot find attribute `serde` in this scope
  --> concrete_proto/src/server.rs:39:7
   |
39 |     #[serde(skip)]
   |       ^^^^^

error: aborting due to 16 previous errors

error: could not compile `concrete_proto`.

To learn more, run the command again with --verbose.
```

**Comment by Yuhanun** [comment]
Any idea?

**Comment by vklein52** [comment]
Running into the same issue myself

**Comment by Nerixyz** [comment]
I had the same issue, this is how I "fixed" it.

In your `Cargo.toml` you need to add this:

```toml
[features]
default = ["with-serde"]
with-serde = ["protobuf/with-serde"]

[dependencies]
# note the derive feature
serde = {version = "1", features = ["derive"] }
```

Your `build.rs` should look something like this:
```rust
// or the other one
use protobuf_codegen_pure::{Customize, Codegen};
use std::fs::{read_to_string, write};

fn main() {
    println!("cargo:rerun-if-changed=protos/FILE_NAME.proto");

    Codegen::new()
        .out_dir("src/protos")
        .input("protos/FILE_NAME.proto")
        .include("protos")
        .customize(Customize {
            serde_derive: Some(true),
            ..Default::default()
        })
        .run()
        .expect("protobuf failed");
    let contents = read_to_string("src/protos/FILE_NAME.rs").expect("file doesn't exist");
    // the import needs to be above the first macro
    let (header, body) = contents.split_at(contents.find("#[").expect("could not find pattern '#['"));
    let contents = format!("{}{}{}", header, "use serde::{Deserialize, Serialize};\n", body);
    write("src/protos/FILE_NAME.rs", contents).expect("could not write file");
}
```

**Comment by stepancheg** [comment]
I supposed this is fixed in 2.21.0. Please report if not.
